main.js

```diff
@@ -0,0 +1,313 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+import {
+  encode as base58encoder,
+  decode as base58decoder
+} from 'base58-universal';
+import {
+  getRandomBytes,
+  bytesToHex,
+  bytesFromHex
+} from './util.js';
+
+function _calcOptionsBitLength({
+  defaultLength,
+  // TODO: allow any bit length
+  minLength = 8,
+  maxLength = Infinity,
+  bitLength
+}) {
+  if(bitLength === undefined) {
+    return defaultLength;
+  }
+  // TODO: allow any bit length
+  if(bitLength % 8 !== 0) {
+    throw new Error('Bit length must be a multiple of 8.');
+  }
+  if(bitLength < minLength) {
+    throw new Error(`Minimum bit length is ${minLength}.`);
+  }
+  if(bitLength > maxLength) {
+    throw new Error(`Maximum bit length is ${maxLength}.`);
+  }
+  return bitLength;
+}
+
+function _calcDataBitLength({
+  bitLength,
+  maxLength
+}) {
+  if(maxLength === 0) {
+    return bitLength;
+  }
+  if(bitLength > maxLength) {
+    throw new Error(
+      `Input length greater than ${maxLength} bits.`);
+  }
+  return maxLength;
+}
+
+function _bytesWithBitLength({
+  bytes,
+  bitLength
+}) {
+  if(bitLength === 0) {
+    return bytes;
+  }
+  const length = bytes.length * 8;
+  if(length === bitLength) {
+    return bytes;
+  }
+  if(length < bitLength) {
+    // pad start
+    const data = new Uint8Array(bitLength / 8);
+    data.set(bytes, data.length - bytes.length);
+    return data;
+  }
+  // trim start, ensure trimmed data is zero
+  const start = (length - bitLength) / 8;
+  if(bytes.subarray(0, start).some(d => d !== 0)) {
+    throw new Error(
+      `Data length greater than ${bitLength} bits.`);
+  }
+  return bytes.subarray(start);
+}
+
+const _log2_16 = 4;
+function _base16Encoder({bytes, idEncoder}) {
+  let encoded = bytesToHex(bytes);
+  if(idEncoder.encoding === 'base16upper') {
+    encoded = encoded.toUpperCase();
+  }
+  if(idEncoder.fixedLength) {
+    const fixedBitLength = _calcDataBitLength({
+      bitLength: bytes.length * 8,
+      maxLength: idEncoder.fixedBitLength
+    });
+    const wantLength = Math.ceil(fixedBitLength / _log2_16);
+    // pad start with 0s
+    return encoded.padStart(wantLength, '0');
+  } else {
+    return encoded;
+  }
+}
+
+const _log2_58 = Math.log2(58);
+function _base58Encoder({bytes, idEncoder}) {
+  const encoded = base58encoder(bytes);
+  if(idEncoder.fixedLength) {
+    const fixedBitLength = _calcDataBitLength({
+      bitLength: bytes.length * 8,
+      maxLength: idEncoder.fixedBitLength
+    });
+    const wantLength = Math.ceil(fixedBitLength / _log2_58);
+    // pad start with 0s (encoded as '1's)
+    return encoded.padStart(wantLength, '1');
+  } else {
+    return encoded;
+  }
+}
+
+export class IdGenerator {
+  /**
+   * Creates a new IdGenerator instance.
+   *
+   * An IdGenerator generates an array of id bytes.
+   *
+   * @param {Number} [bitLength] - number of bits to generate. May also be an
```

Comment:
might as well enable jsdoc linting rules since you wrote some jsdocs.  Spoiler: This is not conforming.

Reply:
I left that tool out because it complained too much. ;-)  I'll check it out again.